src/sync/sync_session.cpp

```diff
@@ -451,42 +475,64 @@ bool SyncSession::can_wait_for_network_completion() const
     return m_state == &State::active || m_state == &State::dying;
 }
 
-void SyncSession::wait_for_upload_completion(std::function<void()> callback)
+void SyncSession::wait_for_network_completion(std::function<void()> callback, NetworkMode mode)
 {
-    // FIXME: If the session is waiting for a token, the `wait_for_upload` should be deferred until the `bind()`
+    // FIXME: If the session is waiting for a token, the `wait_for_X` should be deferred until the `bind()`
     // instead of just calling the callback immediately.
     REALM_ASSERT(shared_from_this());
-    auto thread = std::thread([this, callback=std::move(callback), self=shared_from_this()]() {
-        {
-            std::unique_lock<std::mutex> lock(m_state_mutex);
-            if (can_wait_for_network_completion()) {
-                REALM_ASSERT(m_session);
-                m_session->wait_for_upload_complete_or_client_stopped();
+    std::unique_lock<std::mutex> lock(m_state_mutex);
+    switch (mode) {
+        case NetworkMode::Upload:
+            ++m_pending_upload_threads;
+            break;
+        case NetworkMode::Download:
+            ++m_pending_download_threads;
+            break;
+    }
+    auto thread = std::thread([this, callback=std::move(callback), self=shared_from_this(), mode]() {
+        std::unique_lock<std::mutex> lock(m_state_mutex);
+        if (can_wait_for_network_completion()) {
```

Comment:
The lock can't be released until the if statement's predicate function has been called. However, I've slightly rearranged things to make my intent clearer.